src/native/corehost/hostmisc/utils.cpp

```diff
@@ -348,14 +348,25 @@ bool try_stou(const pal::string_t& str, unsigned* num)
     return true;
 }
 
-pal::string_t get_dotnet_root_env_var_name()
+bool get_dotnet_root_from_env(pal::string_t* dotnet_root_env_var_name, pal::string_t* recv)
 {
+    *dotnet_root_env_var_name = _X("DOTNET_ROOT_");
+    dotnet_root_env_var_name->append(to_upper(get_arch()));
+    if (pal::getenv(dotnet_root_env_var_name->c_str(), recv))
+        return pal::file_exists(*recv);
+
+#if defined(WIN32)
     if (pal::is_running_in_wow64())
     {
-        return pal::string_t(_X("DOTNET_ROOT(x86)"));
+        *dotnet_root_env_var_name = _X("DOTNET_ROOT(x86)");
+        return get_file_path_from_env(dotnet_root_env_var_name->c_str(), recv);
```

Comment:
We don't fall back to `DOTNET_ROOT` if `DOTNET_ROOOT(x86)` isn't set?

Reply:
Currently we don't; I'm not sure if there's any reason to not do it.

Reply:
Good catch @elinor-fung. We need to fix this.

We need to correctly support this behavior on x86:
- If `DOTNET_ROOT_x86` is defined, use that
- Otherwise fallback to `DOTNET_ROOT` (just like any other platform)

Now since we also have to support `DOTNET_ROOT(x86)` I think the order will have to be:
- `DOTNET_ROOT_x86`
- `DOTNET_ROOT(x86)`
- `DOTNET_ROOT`

We should add tests covering this as well.
It will have a observable change effect: On Windows x86 we never looked at the `DOTNET_ROOT`, now we will. This should not be a breaking change though. It should really only change errors in some cases and in very rare cases things will work where they didn't before.

The "good" news is that this is only affecting `apphost`/`nethost`/`comhost`/`ijwhost` - which are always shipped with the app. So existing apps are not going to be affected by installing new version on the machine. Only rebuilt apps will be affected.